File: Python3/find_mode_in_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findMode(self, root: TreeNode) -> List[int]:        
        def dfs(root):
            if root is None: return []

            s = [root.val]
            if root.left:
                s += dfs(root.left)
            if root.right:
                s += dfs(root.right)
                
            return s
        
        elements = dfs(root)
        max_frequency = 0

        hashmap = {}
        for val in elements:
            if val in hashmap:
                hashmap[val] += 1
            else:
                hashmap[val] = 1
            max_frequency = max(max_frequency, hashmap[val])

        return [k for k, v in hashmap.items() if v == max_frequency]

File: C++/find_mode_in_binary_search_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, map<int, int>& hashmap, int& mx) {
        if(!root) return;
        mx = max(mx, ++hashmap[root->val]);
        dfs(root->left, hashmap, mx);
        dfs(root->right, hashmap, mx);
    }
                    
    vector<int> findMode(TreeNode* root) {
        vector<int> result;
        int max_frequency = 0;
        map<int, int> hashmap;
        dfs(root, hashmap, max_frequency);
             
        for(auto e : hashmap)
            if(e.second == max_frequency) result.push_back(e.first);
        
        return result;
    }
};
